shared/src/pool_cache.rs

```diff
@@ -0,0 +1,461 @@
+use crate::{
+    current_block::{self, CurrentBlockStream},
+    maintenance::Maintaining,
+    pool_fetching::{Pool, PoolFetching},
+};
+use anyhow::{Context, Result};
+use ethcontract::BlockNumber;
+use lru::LruCache;
+use model::TokenPair;
+use std::{
+    collections::{BTreeMap, HashMap, HashSet},
+    num::NonZeroU64,
+    sync::Mutex,
+};
+
+/// Caching pool fetcher
+///
+/// Caches all requests for a specific number of blocks and automatically updates the N most
+/// recently used pools automatically when a new block arrives.
+pub struct PoolCache {
+    mutexed: Mutex<Mutexed>,
+    number_of_blocks_to_cache: NonZeroU64,
+    inner: Box<dyn PoolFetching>,
+    block_stream: CurrentBlockStream,
+}
+
+// Design:
+// The design of this module is driven by the need to always return pools quickly so that end users
+// going through the api do not have to wait longer than necessary:
+// - The mutex is never held while waiting on an async operation (getting pools from the node).
+// - Automatically updating the cache is decoupled from normal pool fetches.
+// A result of this is that it is possible that the same uncached pair is requested multiple times
+// simultaneously and some work is wasted. This is unlikely to happen in practice and the pool is
+// going to be cached the next time it is needed.
+// When pools are requested we mark all those pools as recently used which potentially evicts other
+// pairs from the pair lru cache. Cache misses are fetched and inserted into the cache.
+// Then when the automatic update runs the next time, we request and cache all recently used pairs.
+// For the order book we only care about the "recent" state of the pools. So we can return any
```

Comment:
nanonit: I really like the design consideration docs, just for "who uses it how" we could maybe be a bit more generic ("some consumers only care about the recent state"). Otherwise I feel like this comment won't age well with the rest of the codebase changing. But also no big deal